src/rules/importBlacklistRule.ts

```diff
@@ -54,9 +89,140 @@ export class Rule extends Lint.Rules.AbstractRule {
 }
 
 function walk(ctx: Lint.WalkContext<string[]>) {
+    type Options = Array<string | { [moduleName: string]: string[] }>;
+    interface BannedImports {
+        [moduleName: string]: true | Set<string>;
+    }
+
+    // Merge/normalize options.
+    // E.g., ["a", { "b": ["c"], "d": ["e"] }, "f", { "f": ["g"] }]
+    // becomes { "a": true, "b": Set(["c"]), "d": Set(["e"]), "f": true }.
+    const bannedImports = (ctx.options as Options).reduce<BannedImports>(
+        (acc, it) => {
+            if (typeof it === "string") {
+                acc[it] = true;
+            } else {
+                Object.keys(it).forEach((moduleName) => {
+                    if (acc[moduleName] instanceof Set) {
+                        it[moduleName].forEach((bannedImport) => {
+                            (acc[moduleName] as Set<string>).add(bannedImport);
+                        });
+                    } else if (acc[moduleName] !== true) {
+                        acc[moduleName] = new Set(it[moduleName]);
+                    }
+                });
+            }
+            return acc;
+        },
+        {},
+    );
+
     for (const name of findImports(ctx.sourceFile, ImportKind.All)) {
-        if (ctx.options.indexOf(name.text) !== -1) {
-            ctx.addFailure(name.getStart(ctx.sourceFile) + 1, name.end - 1, Rule.FAILURE_STRING);
+        // TODO #3963: Resolve/normalize relative file imports to a canonical path?
+        const importedModule = name.text;
+        const bansForModule = bannedImports[importedModule];
+
+        // Check if at least some imports from this module are banned.
+        if (hasOwnProperty(bannedImports, importedModule)) {
```

Comment:
We couldn't previously switch to that, because it would've gotten tripped up by properties added to `Object.prototype`. But I've switched how `bannedImports` is created and refactored this check accordingly.